### PR TITLE
Tenderdash threshold signing

### PR DESCRIPTION
**Pull request by QuantumExplorer**
This pull request incorporates threshold signing into Tenderdash.

There are many changes in this PR:
*Creation of threshold recovered signatures: these signatures can verify that the minimum threshold of members of a validator set have signed a block. Currently in the commit I am keeping all votes, but these will most likely be removed in the future.
*In this PR I also created a StateId for each block that is signed. The stateId was being threshold signed and this signature was proof that the state was merkleized to a certain hash at a given height. I did it this way to keep the BlockId unchanged as to not break the current system. This introduced quite a few issues in terms of pBFT such as what would happen if the StateId is threshold signed but not the BlockId (in case of some byzantine validators). In a future PR I will incorporate the StateId information into the blockId. Doing so will lead to slightly large light client proofs, but most likely only about 32 extra bytes which is manageable.
*Removal of BFT time (needed so that all vote signatures are the same for threshold recovery), time was added to the proposer. Validators will vote nil in a prevote phase if the proposed time is more than 30s from their clock time.
*Changes to the validator set: Validators are now referenced by a proTxHash (meaning provider transaction hash), this is the same reference that they have in members of a quorum in Dash Core. Validators can now have their keys changed as a result. Validator sets now are more than a list of validator changes. A validator set change also includes a changed threshold public key.

### CODE REVIEW

File: proto/tendermint/abci/types.proto [hunk 1]
```diff
@@ -168,6 +169,7 @@ message ResponseInfo {
 
   int64 last_block_height   = 4;
   bytes last_block_app_hash = 5;
+  uint32 last_core_chain_locked_height   = 100;
```
Comment (Cofresi) [inline proto/tendermint/abci/types.proto:172]: According to this JIRA task we need to obtain masternode info on start and pass it to Tenderdash with the ResponseInfo response.
https://dashpay.atlassian.net/secure/RapidBoard.jspa?rapidView=101&modal=detail&selectedIssue=DP-2938
So I guess we need to be able to respond with the same fields for the individual node as are in the `ValidatorUpdate`? Specifically `proTxHash`, `pubKey` and `power`? In that case there should be an additional `validator_update` variable here in the ResponseInfo message.

**Review by Cofresi — CHANGES_REQUESTED** [review]

File: proto/tendermint/abci/types.proto (hunk 1 above)
Comment (QuantumExplorer) [reply proto/tendermint/abci/types.proto:172]: @Cofresi That is not correct. I might have given a slight incorrect demand to the platform team before. Basically the signer client (core) needs to be respond to the query for the proTxHash and the publicKey. Platform needs to do nothing. in that regard.

Comment (Cofresi) [reply proto/tendermint/abci/types.proto:172]: Oh, that is good news! Happy to close the task mentioned above then on Monday

File: Makefile [hunk 2]
```diff
@@ -273,7 +273,7 @@ endif
 
 # Run a nodejs tool to test endpoints against a localnet
 # The command takes care of starting and stopping the network
-# prerequisits: build-contract-tests-hooks build-linux
+# prerequisites: build-contract-tests-hooks build-linux
```
Comment (shumkov) [inline Makefile:276]: Take it out

File: types/block.go [hunk 3]
```diff
@@ -347,8 +368,8 @@ type Header struct {
 	LastResultsHash tmbytes.HexBytes `json:"last_results_hash"`
 
 	// consensus info
-	EvidenceHash    tmbytes.HexBytes `json:"evidence_hash"`    // evidence included in the block
-	ProposerAddress Address          `json:"proposer_address"` // original proposer of the block
+	EvidenceHash      tmbytes.HexBytes `json:"evidence_hash"`        // evidence included in the block
+	ProposerProTxHash ProTxHash        `json:"proposer_pro_tx_hash"` // original proposer of the block
```
Comment (shumkov) [inline types/block.go:372]: `ProTx` -> `CoreProTx`

File: types/block.go [hunk 4]
```diff
@@ -1271,3 +1383,72 @@ func BlockIDFromProto(bID *tmproto.BlockID) (*BlockID, error) {
...
+}
+
+// IsZero returns true if this is the StateID of a nil block.
+func (stateID StateID) IsZero() bool {
+	return len(stateID.LastAppHash) == 0
+}
+
+// IsComplete returns true if this is a valid StateID of a non-nil block.
+func (stateID StateID) IsComplete() bool {
+	return len(stateID.LastAppHash) == tmhash.Size
```
Comment (shumkov) [inline types/block.go:1420]: Introduce an app hash size option to tenderdash config.

File: config/toml.go [hunk 5]
```diff
@@ -530,10 +531,15 @@ var testGenesisFmt = `{
...
         "value":"F5BjXeh0DppqaxX7a3LzoWr6CXPZcZeba6VHYdbiUCxQ23b00mFD8FRZpCz9Ug1E"
       },
-      "power": "10",
-      "name": ""
+      "power": "100",
+      "name": "",
+      "pro_tx_hash": "51BF39CC1F41B9FC63DFA5B1EDF3F0CA3AD5CAFAE4B12B4FE9263B08BB50C45F"
     }
   ],
+  "threshold_public_key": {
```
Comment (shumkov) [inline config/toml.go:539]: Maybe it makes sense to add a `validators_` prefix because they depend on each other.

File: consensus/state.go [hunk 6]
```diff
@@ -1734,13 +1777,22 @@ func (cs *State) defaultSetProposal(proposal *types.Proposal) error {
...
-	) {
-		return ErrInvalidProposalSignature
+	proposalBlockSignBytes := types.ProposalBlockSignBytes(cs.state.ChainID, p)
+	proposer := cs.Validators.GetProposer()
+
+	if !proposer.PubKey.VerifySignature(proposalBlockSignBytes, proposal.Signature) {
+		return fmt.Errorf("error proposer %X verifying proposal signature %X at height %d with key %X blockSignBytes %X",
+			proposer.ProTxHash, proposal.Signature, proposal.Height, proposer.PubKey.Bytes(), proposalBlockSignBytes)
 	}
+	// } else if proposal.Height == 9 {
```
Comment (shumkov) [inline consensus/state.go:1793]: Remove comment

File: crypto/bls12381/bls12381.go [hunk 7]
```diff
@@ -124,12 +136,194 @@ func genPrivKey(rand io.Reader) PrivKey {
...
 	seed := crypto.Sha256(secret) // Not Ripemd160 because we want 32 bytes.
+	privKey, err := bls.PrivateKeyFromSeed(seed)
+	if err != nil {
+		panic(err)
+	}
+	return PrivKey(privKey.Serialize())
+}
 
-	privateKey, err := bls.PrivateKeyFromSeed(seed)
+func CreatePrivLLMQDataDefaultThreshold(members int) ([]crypto.PrivKey, []crypto.ProTxHash, crypto.PubKey) {
```
Comment (shumkov) [inline crypto/bls12381/bls12381.go:146]: Move these functions to helpers module.

File: docs/architecture/adr-025-commit.md [hunk 8]
```diff
@@ -15,7 +15,7 @@ type Commit struct {
 }
 
 type Vote struct {
-    ValidatorAddress Address   `json:"validator_address"`
+    ValidatorProTxHash Address `json:"validator_pro_tx_hash"`
```
Comment (shumkov) [inline docs/architecture/adr-025-commit.md:18]: Bring back

File: privval/file.go [hunk 9]
```diff
@@ -46,9 +46,12 @@ func voteToStep(vote *tmproto.Vote) int8 {
 
 // FilePVKey stores the immutable part of PrivValidator.
 type FilePVKey struct {
-	Address types.Address  `json:"address"`
-	PubKey  crypto.PubKey  `json:"pub_key"`
-	PrivKey crypto.PrivKey `json:"priv_key"`
+	Address            types.Address    `json:"address"`
+	PubKey             crypto.PubKey    `json:"pub_key"`
+	PrivKey            crypto.PrivKey   `json:"priv_key"`
+	NextPrivKeys       []crypto.PrivKey `json:"next_priv_key,omitempty"`
```
Comment (shumkov) [inline privval/file.go:52]: Looks a bit strange, I need to think about this more.

File: proto/tendermint/abci/types.proto [hunk 10]
```diff
@@ -62,9 +62,10 @@ message RequestInitChain {
       [(gogoproto.nullable) = false, (gogoproto.stdtime) = true];
   string                   chain_id         = 2;
   ConsensusParams          consensus_params = 3;
-  repeated ValidatorUpdate validators       = 4 [(gogoproto.nullable) = false];
+  //repeated ValidatorUpdate validators       = 4 [(gogoproto.nullable) = false];
   bytes                    app_state_bytes  = 5;
   int64                    initial_height   = 6;
+  ValidatorSetUpdate validator_set = 100 [(gogoproto.nullable) = false];
```
Comment (shumkov) [inline proto/tendermint/abci/types.proto:68]: Should we use this gap 100?

File: proto/tendermint/privval/types.proto [hunk 11]
```diff
@@ -72,5 +83,7 @@ message Message {
     SignedProposalResponse signed_proposal_response = 6;
     PingRequest            ping_request             = 7;
     PingResponse           ping_response            = 8;
+    ProTxHashRequest          pro_tx_hash_request          = 9;
```
Comment (shumkov) [inline proto/tendermint/privval/types.proto:86]: If we keep gaps we need to have it here as well

File: proto/tendermint/types/canonical.proto [hunk 12]
```diff
@@ -32,6 +36,11 @@ message CanonicalVote {
   sfixed64                  height    = 2;  // canonicalization requires fixed size encoding here
   sfixed64                  round     = 3;  // canonicalization requires fixed size encoding here
   CanonicalBlockID          block_id  = 4 [(gogoproto.customname) = "BlockID"];
-  google.protobuf.Timestamp timestamp = 5 [(gogoproto.nullable) = false, (gogoproto.stdtime) = true];
+  CanonicalStateID          state_id  = 7 [(gogoproto.customname) = "StateID"];
```
Comment (shumkov) [inline proto/tendermint/types/canonical.proto:39]: Gap if needed

File: proto/tendermint/types/types.proto [hunk 13]
```diff
@@ -97,38 +111,40 @@ message Vote {
...
   BlockID       block_id = 4
       [(gogoproto.nullable) = false, (gogoproto.customname) = "BlockID"];  // zero if vote is nil.
-  google.protobuf.Timestamp timestamp = 5
-      [(gogoproto.nullable) = false, (gogoproto.stdtime) = true];
-  bytes validator_address = 6;
+  StateID       state_id = 9 [(gogoproto.customname) = "StateID", (gogoproto.nullable) = false];
+  bytes validator_pro_tx_hash = 6;
   int32 validator_index   = 7;
-  bytes signature         = 8;
+  bytes block_signature         = 8;
```
Comment (shumkov) [inline proto/tendermint/types/types.proto:117]: Gap if needed

File: rpc/client/interface.go [hunk 14]
```diff
@@ -68,7 +68,8 @@ type SignClient interface {
 	BlockByHash(ctx context.Context, hash []byte) (*ctypes.ResultBlock, error)
 	BlockResults(ctx context.Context, height *int64) (*ctypes.ResultBlockResults, error)
 	Commit(ctx context.Context, height *int64) (*ctypes.ResultCommit, error)
-	Validators(ctx context.Context, height *int64, page, perPage *int) (*ctypes.ResultValidators, error)
+	Validators(ctx context.Context, height *int64, page, perPage *int,
```
Comment (shumkov) [inline rpc/client/interface.go:71]: Probably it makes sense to add an extra endpoint to get the public key.

File: state/execution.go [hunk 15]
```diff
@@ -374,6 +431,17 @@ func getBeginBlockValidatorInfo(block *types.Block, store Store,
 	}
 }
 
+func validateValidatorSetUpdate(abciValidatorSetUpdate *abci.ValidatorSetUpdate, params tmproto.ValidatorParams) error {
+	// if there was no update return no error
+	if abciValidatorSetUpdate == nil {
+		return nil
+	}
+	if len(abciValidatorSetUpdate.ValidatorUpdates) != 0 && abciValidatorSetUpdate.ThresholdPublicKey.Sum == nil {
```
Comment (shumkov) [inline state/execution.go:439]: Let's do the opposite check.

File: state/helpers_test.go [hunk 16]
```diff
@@ -129,87 +133,73 @@ func makeState(nVals, height int) (sm.State, dbm.DB, map[string]types.PrivValida
...
-		abciResponses.EndBlock = &abci.ResponseEndBlock{
-			ValidatorUpdates: []abci.ValidatorUpdate{
-				types.TM2PB.NewValidatorUpdate(val.PubKey, power),
-			},
-		}
-	}
-
-	return block.Header, types.BlockID{Hash: block.Hash(), PartSetHeader: types.PartSetHeader{}}, abciResponses
-}
+// func makeHeaderPartsResponsesValPowerChange(
```
Comment (shumkov) [inline state/helpers_test.go:168]: Remove commented code

File: test/maverick/consensus/state.go [hunk 17]
```diff
@@ -1811,43 +1831,50 @@ func (cs *State) signVote(
...
-	} else if cs.ProposalBlock != nil {
-		minVoteTime = cs.ProposalBlock.Time.Add(timeIota)
-	}
-
-	if now.After(minVoteTime) {
-		return now
-	}
-	return minVoteTime
-}
+// func (cs *State) voteTime() time.Time {
```
Comment (shumkov) [inline test/maverick/consensus/state.go:1860]: Remove comment

File: evidence/verify.go [hunk 18]
```diff
@@ -194,11 +196,9 @@ func VerifyDuplicateVote(e *types.DuplicateVoteEvidence, chainID string, valSet
...
 	}
 
-	// pubkey must match address (this should already be true, sanity check)
-	addr := e.VoteA.ValidatorAddress
-	if !bytes.Equal(pubKey.Address(), addr) {
-		return fmt.Errorf("address (%X) doesn't match pubkey (%v - %X)",
-			addr, pubKey, pubKey.Address())
+	// proTxHash must match address (this should already be true, sanity check)
+	if !bytes.Equal(proTxHash, e.VoteA.ValidatorProTxHash) {
+		return fmt.Errorf("proTxHash (%X) doesn't match pubkey (%v)", e.VoteA.ValidatorProTxHash, proTxHash)
```
Comment (Cofresi) [inline evidence/verify.go:201]:
```suggestion
		return fmt.Errorf("proTxHash (%X) doesn't match ValidatorProTxHash(%v)", e.VoteA.ValidatorProTxHash, proTxHash)
```

File: evidence/verify.go [hunk 19]
```diff
@@ -194,11 +196,9 @@ func VerifyDuplicateVote(e *types.DuplicateVoteEvidence, chainID string, valSet
 		)
 	}
 
-	// pubkey must match address (this should already be true, sanity check)
-	addr := e.VoteA.ValidatorAddress
-	if !bytes.Equal(pubKey.Address(), addr) {
-		return fmt.Errorf("address (%X) doesn't match pubkey (%v - %X)",
-			addr, pubKey, pubKey.Address())
+	// proTxHash must match address (this should already be true, sanity check)
```
Comment (Cofresi) [inline evidence/verify.go:199]:
```suggestion
	// proTxHash must match ValidatorProTxHash(this should already be true, sanity check)
```

**Review by Cofresi — CHANGES_REQUESTED** [review]

**Pull request merged**